linalg/constraints.hpp

```diff
@@ -0,0 +1,404 @@
+    EliminationProjection and Z_P is the identity on the eliminated dofs. */
+class EliminationSolver : public ConstrainedSolver
+{
+public:
+   /** @brief Constructor, with explicit splitting into primary/secondary dofs.
+
+       This constructor uses a single elimination block (per processor), which
+       provides the most general algorithm but is also not scalable
+
+       The secondary_dofs are eliminated from the system in this algorithm,
+       as they can be written in terms of the primary_dofs. */
+   EliminationSolver(HypreParMatrix& A, SparseMatrix& B,
+                     Array<int>& primary_dofs,
+                     Array<int>& secondary_dofs);
+
+   /** @brief Constructor, elimination is by blocks.
+
+       The nonzeros in B are assumed to be in disjoint rows and columns; the
+       rows are identified with the lagrange_rowstarts array, the secondary
+       dofs are assumed to be the first nonzeros in the rows. */
+   EliminationSolver(HypreParMatrix& A, SparseMatrix& B,
+                     Array<int>& lagrange_rowstarts);
+
+   ~EliminationSolver();
+
+   void PrimalMult(const Vector& x, Vector& y) const override;
+
+protected:
+   /// Internal utility routine; assembles eliminated matrix explicitly
+   void BuildExplicitOperator();
+
+   /// Build preconditioner for eliminated system
+   virtual Solver* BuildPreconditioner() const = 0;
+   /// Select krylov solver for eliminated system
+   virtual IterativeSolver* BuildKrylov() const = 0;
+
+   HypreParMatrix& hA;
+   Array<Eliminator*> eliminators;
+   EliminationProjection * projector;
+   HypreParMatrix * h_explicit_operator;
+   mutable Solver* prec;
+};
+
+
+/** EliminationSolver using CG and HypreBoomerAMG */
+class EliminationCGSolver : public EliminationSolver
+{
+public:
+   EliminationCGSolver(HypreParMatrix& A, SparseMatrix& B,
+                       Array<int>& lagrange_rowstarts,
+                       int dimension_=0, bool reorder_=false) :
+      EliminationSolver(A, B, lagrange_rowstarts),
+      dimension(dimension_), reorder(reorder_)
+   { BuildExplicitOperator(); }
+
+protected:
+   virtual Solver* BuildPreconditioner() const override
+   {
+      HypreBoomerAMG * h_prec = new HypreBoomerAMG(*h_explicit_operator);
+      h_prec->SetPrintLevel(0);
+      if (dimension > 0) { h_prec->SetSystemsOptions(dimension, reorder); }
+      return h_prec;
+   }
+
+   virtual IterativeSolver* BuildKrylov() const override
+   { return new CGSolver(GetComm()); }
+
+private:
+   int dimension;
+   bool reorder;
+};
+
+
+/** @brief Solve constrained system with penalty method; see ConstrainedSolver.
+
+    Only approximates the solution, better approximation with higher penalty,
+    but with higher penalty the preconditioner is less effective. */
+class PenaltyConstrainedSolver : public ConstrainedSolver
+{
+public:
+   PenaltyConstrainedSolver(HypreParMatrix& A, SparseMatrix& B,
+                            double penalty_);
```

Comment:
While I think it's goes beyond the scope of this initial PR, do you think it would be difficult to extend it to use different penalties for each constraint? In other words, the penalty parameter would be a vector instead of a double.

Reply:
That should be straightforward.